### PR TITLE
Interval support for OFFSET PRECEDING/FOLLOWING

### PR DESCRIPTION
**Pull request by marregui**
…ession

This commit message will be edited later on.

## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [ ] Touched code is covered by tests
 - [ ] Documentation has been updated if necessary
 - [ ] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [ ] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by marregui** [comment]
First pass, I still need to add appropriate junit tests, documentation, and go twice over the code, as my implementation did not touch `FrameBound` or `ComputeFrameBoundary`.

File: sql/src/test/java/io/crate/integrationtests/WindowFunctionsIntervalIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Arrays;
+
+import static io.crate.testing.TestingHelpers.printedTable;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.is;
+
+public class WindowFunctionsIntervalIntegrationTest extends SQLTransportIntegrationTest {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/WindowFunctionsIntervalIntegrationTest.java]: Can we cover the functionality using unit tests instead of integration tests? They're an order of magnitude faster

File: docs/general/builtins/window-functions.rst [hunk 2]
```diff
@@ -106,13 +106,17 @@ In ``ROWS`` mode ``CURRENT_ROW`` means the current row.
...
-after the current row. In ``RANGE`` mode, the use of a custome ``offset``
+after the current row. In ``RANGE`` mode, the use of a custom ``offset``
 option requires that there is exactly one ``ORDER BY`` column in the window
 definition. The frame contains those rows whose ordering column value is no
 more than ``offset`` minus (for PRECEDING) or plus (for FOLLOWING) the current
 row's ordering column value. In this case the data type of the ``offset``
 expression must have the same type of the ordering column.
 
+.. note::
+    When the ordering column is of type :ref:`timestamp <timestamp_data_type>`,
```
Comment (mfussenegger) [inline docs/general/builtins/window-functions.rst]: Maybe it would make more sense to re-phrase the last sentence of the previous paragaph. To explain that the offset value is substracted ot added to the ordering value, and so only type combinations which support add/substract are supported.

File: sql/src/test/java/io/crate/integrationtests/WindowFunctionsIntervalIntegrationTest.java (hunk 1 above)
Comment (marregui) [reply sql/src/test/java/io/crate/integrationtests/WindowFunctionsIntervalIntegrationTest.java]: yes

File: sql/src/test/java/io/crate/execution/engine/window/WindowDefinitionTest.java [hunk 3]
```diff
@@ -58,6 +58,66 @@ public void prepare() throws IOException {
...
+            "select id, avg(date::long) over(twenty_sec_following)" +
+            "  from doc.users" +
+            "  window twenty_sec_following as (rows between current row and '20 seconds'::interval following)" +
+            "  order by id");
+    }
+
+    @Test
+    public void test_offset_following_does_not_support_negative_values_in_rows_mode() {
+        expectedException.expectMessage(containsString("In ROWS mode the offset must be a non-null, non-negative number"));
+        e.analyze(
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/execution/engine/window/WindowDefinitionTest.java]: These kind of semantic / error handling tests don't really have anything to do with adding interval support.

Can we revert this from this PR and if they're useful to add, do that in a separate PR? (I thought we cover this already elsewhere)

File: sql/src/test/java/io/crate/execution/engine/window/WindowDefinitionTest.java [hunk 4]
```diff
@@ -58,6 +58,66 @@ public void prepare() throws IOException {
...
+    @Test
+    public void test_offset_following_supports_interval_in_range_mode() {
+        e.analyze(
+            "select id," +
+            "       avg(ints) over(thirty_secs_following)," +
+            "       avg(date::long) over(thirty_secs_following) as \"avg date 30 seconds following (range mode)\"" +
+            "  from doc.users" +
+            "  window thirty_secs_following as (order by date range between current row and '30 seconds'::interval following)" +
+            "  order by id");
+    }
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/execution/engine/window/WindowDefinitionTest.java]: These tests don't do much except making sure that the query can be analysed somehow.

When I asked to transform the integration tests to unit tests I meant to utilize `AbstractWindowFunctionTest` so that the function execution / windowing logic is executed and tested.

File: sql/src/test/java/io/crate/execution/engine/window/WindowDefinitionTest.java (hunk 3 above)
Comment (marregui) [reply sql/src/test/java/io/crate/execution/engine/window/WindowDefinitionTest.java]: remove them, ok

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 5]
```diff
@@ -314,13 +315,15 @@ private WindowDefinition getWindowDefinition(Optional<Window> maybeWindow,
         WindowFrameDefinition windowFrameDefinition = WindowDefinition.RANGE_UNBOUNDED_PRECEDING_CURRENT_ROW;
         if (window.getWindowFrame().isPresent()) {
             WindowFrame windowFrame = window.getWindowFrame().get();
-            validateFrame(window, windowFrame);
+            WindowFrameBoundsValidator.validateFrame(window, windowFrame);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: I think the changes here should also no longer be necessary?

Comment (marregui) [reply sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: there are two parts to validation, one covers the invariant that is independent from the bound value, the other is a post validation once the values are bound. These two syntax checks go together and I thought the ..validator package contained classes related to validation. I think the code reads well.

Comment (mfussenegger) [reply sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: My point is that I don't understand why this change is necessary.

Given that the code has been moved it is also difficult to see the diff. It would be good to always split formating or code structure changes from behavior changes, so that it is easier to see the actual change.

Doing "bound value" validation in the analyzer sounds wrong to me. Queries have to fail with a appropriate error message in any case, so we have to assume that the analyzer *does not* have access to bound values. So the validation needs to happen later (planner or execution layer).

Comment (marregui) [reply sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: I agree, reverting.

File: docs/appendices/release-notes/unreleased.rst [hunk 6]
```diff
@@ -65,6 +65,10 @@ Deprecations
 Changes
 =======
 
+- Added support for type interval in ``offset PRECEDING/FOLLOWING``. When the
```
Comment (mfussenegger) [inline docs/appendices/release-notes/unreleased.rst]: Maybe clarify that this is about window functions.
